### PR TITLE
[release/6.0] fix IsMutuallyAuthenticated on SslStream

### PR DESCRIPTION
**Pull request by wfurt**
This is backport of PR #88488 and PR #79128 and parts of PR #63945. 
It also brings spirit of test-only PR #68009 to get test coverage for TLS 1.3.  

This only covers Windows to minimize the code delta i.e. it does not bring all the changes from PR #63945 to cover Linux & macOS.

## Customer Impact

The property `IsMutuallyAuthenticated` on `SslStream` indicates if mutual TLS authentication is performed with client certificate. Current 6.0 implementation can get confused in several cases, so the value is unreliable for security audits.

## Testing

This brings all the current tests from 8.0 branch.
Customer validated on private bits in production - neither functional, nor perf regression.

## Risk

Medium.
While the change is quite large, it should be specific just to that property i.e. it should not impact TLS handshake or any other I/O on `SslStream`. Since the `IsMutuallyAuthenticated` is already unreliable this should bring it up to 8.0 code base to fix all known cases when it is incorrect. To reduce complexity, this fixes only Windows as macOS & Linux changes from PR #68009 had more significant impact on functionality and flow.

### CODE REVIEW

**Review by rzikm — APPROVED** [review]
LGTM if CI is green

**Comment by carlossanlop** [comment]
> LGTM if CI is green

@wfurt please send an email to Tactics requesting approval and add the `servicing-consider` label. I couldn't find an email yet. We still have time to include this in the November release.

**Comment by karelz** [comment]
@carlossanlop we will bring it in for December. We need to prepare also 7.0 backport - fixing only 6.0 would be weird. And as you see from the delta, it is rather involved change, so I don't want to rush it.

**Comment by carlossanlop** [comment]
Friendly reminder: If you'd like this to be included in the December release, please merge it before Tuesday November 14th EOD (Code Complete).

**Comment by karelz** [comment]
Thanks @carlossanlop we want to get validation on privates before we send it to Tactics. We will miss also December release.

File: src/libraries/Common/src/Interop/Windows/SspiCli/SecuritySafeHandles.cs [hunk 1]
```diff
@@ -320,10 +321,15 @@ public new IntPtr DangerousGetHandle()
...
     {
+#pragma warning disable 0649
+         // This is used only by SslStream but it is included elsewhere
+         public X509Certificate? LocalCertificate;
+ #pragma warning restore 0649
         public SafeFreeCredential_SECURITY() : base() { }
 
         protected override bool ReleaseHandle()
         {
+            LocalCertificate?.Dispose();
```
Comment (stephentoub) [inline src/libraries/Common/src/Interop/Windows/SspiCli/SecuritySafeHandles.cs:332]: Same question as on the 7.0 PR

**Review by stephentoub — APPROVED** [review]

File: src/libraries/Common/src/Interop/Windows/SspiCli/SecuritySafeHandles.cs (hunk 1 above)
Comment (rzikm) [reply src/libraries/Common/src/Interop/Windows/SspiCli/SecuritySafeHandles.cs:332]: Let's keep the disucssion here for both PRs.

I don't see anything in the dispose implementation nor transitively called methods that could throw. The Reset method which is called there is virtual, *but* we explicitly make sure we assign only `X509Certificate2` instances to the field.

Comment (wfurt) [reply src/libraries/Common/src/Interop/Windows/SspiCli/SecuritySafeHandles.cs:332]: This is what current `main` does. If we are concern we may wrap it In `try/catch` or simply re-order the code. In either case my preference would be to keep it consistent across all supported branches to make future back ports easier

File: src/libraries/Common/src/Interop/Windows/SspiCli/SSPIWrapper.cs [hunk 2]
```diff
@@ -426,27 +426,41 @@ private static unsafe int EncryptDecryptHelper(OP op, ISSPIInterface secModule,
...
         }
 
+        public static bool QueryContextAttributes_SECPKG_ATTR_REMOTE_CERT_CONTEXT(ISSPIInterface secModule, SafeDeleteContext securityContext, out SafeFreeCertContext? certContext)
+            => QueryCertContextAttribute(secModule, securityContext, Interop.SspiCli.ContextAttribute.SECPKG_ATTR_REMOTE_CERT_CONTEXT, out certContext);
+
+        public static bool QueryContextAttributes_SECPKG_ATTR_LOCAL_CERT_CONTEXT(ISSPIInterface secModule, SafeDeleteContext securityContext, out SafeFreeCertContext? certContext)
+            => QueryCertContextAttribute(secModule, securityContext, Interop.SspiCli.ContextAttribute.SECPKG_ATTR_LOCAL_CERT_CONTEXT, out certContext);
+
+        public static bool QueryContextAttributes_SECPKG_ATTR_REMOTE_CERT_CHAIN(ISSPIInterface secModule, SafeDeleteContext securityContext, out SafeFreeCertContext? certContext)
+            => QueryCertContextAttribute(secModule, securityContext, Interop.SspiCli.ContextAttribute.SECPKG_ATTR_REMOTE_CERT_CHAIN, out certContext);
```
[the remark this answers is not in the document]
Comment (wfurt) [reply src/libraries/Common/src/Interop/Windows/SspiCli/SSPIWrapper.cs:461]: updated

**Comment by karelz** [comment]
Approved by Tactics (@SteveMCarroll) on 1/9 via email. Adding Servicing-approved label accordingly.

**Pull request merged**